### ISSUE TITLE
How to decide the labels_weight in SoftmaxWithLossW?

### ISSUE THREAD

**Issue by John1231983** [ginobilinie/infantSeg#4]
Hello, you are using labels_weight to handle with unbalanced classification problem in infant segmentation. **Could you tell me how much gain do you have when you use labels_weight scheme?**
Second, your prototxt is below. It has two parameters in softmax loss layer:`weight_labels` and `labels_weight`. I guess `weight_labels`  are label name, i.e 0 for Background, 1 for CSF, 2 for WM, and 3 for GM. And the `labels_weight` is the weight for each class. **Could you tell me how do you decide the values?** Besides, I did not find any SoftmaxWithLossW in caffe. I guess it is your custom layer. If it is possible, **could you share to me the layer implementation**?
```
layer {
  name: "loss"
  type: "SoftmaxWithLossW"
  bottom: "conv7"
  bottom: "dataSeg"
  top: "loss"
  softmax_lossw_param {
    weight_labels: 0
    weight_labels: 1
    weight_labels: 2 
    weight_labels: 3
    labels_weight: 1
    labels_weight: 4
    labels_weight: 8
    labels_weight: 6
  }
  loss_param {
    ignore_label: -1
  }
  softmax_param {
    axis: 1
  }
}
```

**Comment by ginobilinie** [comment]
@John1231983 Using weight loss scheme, the test loss can be decreased from 0.09-0.10 to 0.06-0.07, as for  Dice, it also can help improve 2-5%.
As for how to define the weights, actually, you can estimate according to each type's share over the whole image (e.g., CSF occupies 2%, then you can give a higher weights for CSF). And usually people use it to decrease the impact of background, in my case, I use it for all 3 types.
Of course, I can share that copy of code: https://github.com/ginobilinie/infantSeg/blob/master/softmax_lossweight_layer.cpp
I hope it can help you.

**Comment by John1231983** [comment]
Thanks for sharing it. Could you also share the .hpp file of the layer also?

I see. So, you will measure how many percents each class will contribute in the whole brain. Is it right? Does it measure based on patches or whole brain? Base on patch means we will randomly crop a patch of 32x32x32, then calculate how many percent of BG, CSF, GM, WM, then repeat it about 10.000 times (likes 10.000 patches) and the final percentage is the average of these values. Another option that we will compute the percentage for each class based on whole image/subject. The final percentage is an average of all subject in the training set. Which option did you prefer?

Assume that I have the proportion of each class likes 10% CSF, 20% for GM, 30% for WM and 40% for BG. Based on your suggestion, then the weights for each class are [4, 2, 4/3,1]. Am I right?
